origin_ui/register_origin.go

```diff
@@ -0,0 +1,200 @@
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package origin_ui
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/metrics"
+	nsregistry "github.com/pelicanplatform/pelican/namespace-registry"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	errorResp struct {
+		Error string `json:"error"`
+	}
+
+	keyStatus int
+)
+
+const (
+	noKeyPresent keyStatus = iota
+	keyMismatch
+	keyMatch
+)
+
+func keyIsRegistered(privkey jwk.Key, url string) (keyStatus, error) {
+	keyId := privkey.KeyID()
+	if keyId == "" {
+		return noKeyPresent, errors.New("Provided key is missing a key ID")
+	}
+	key, err := privkey.PublicKey()
+	if err != nil {
+		return noKeyPresent, err
+	}
+
+	req, err := http.NewRequest("GET", url+"/.well-known/issuer.jwks", nil)
+	if err != nil {
+		return noKeyPresent, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	tr := config.GetTransport()
+	client := &http.Client{Transport: tr}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return noKeyPresent, err
+	}
+	defer resp.Body.Close()
+
+	// Check HTTP response -- should be 200, else something went wrong
+	body, _ := io.ReadAll(resp.Body)
+	if resp.StatusCode == 404 || resp.StatusCode == 500 {
+		// TODO: The registry returns a 500 for unregistered namespaces instead of a 404.
+		// It would be better to have an error message in this case but we must instead assume
+		// it's an unregistered namespace.
+		return noKeyPresent, nil
+	} else if resp.StatusCode != 200 {
+		var msg errorResp
+		if err := json.Unmarshal(body, &msg); err != nil {
+			log.Warningln("Failed to unmarshal error message response from namespace registry", err)
+		}
+		if msg.Error != "" {
+			return noKeyPresent, errors.Errorf("Failed to query registry for public key (status code %v): %v", resp.StatusCode, msg.Error)
+		} else {
+			return noKeyPresent, errors.Errorf("Failed to query registry for public key: status code %v", resp.StatusCode)
+		}
+	}
+
+	registrySet, err := jwk.Parse(body)
+	if err != nil {
+		log.Debugln("Failed to parse registry response:", string(body))
+		return noKeyPresent, errors.Wrap(err, "Failed to parse registry response as a JWKS")
+	}
+
+	registryKey, isPresent := registrySet.LookupKeyID(keyId)
+	if !isPresent {
+		return keyMismatch, nil
```

Comment:
Should this instead be `noKeyPresent`?, as `keyMismatch` seems incorrect in this first case.